File: src/org/telegram/messenger/NotificationCenter.ts
/**
 * Telegram Official NotificationCenter Observer System
 * Replicates org.telegram.messenger.NotificationCenter.java
 */

export class NotificationCenter {
  private static instance: NotificationCenter;
  private observers: Map<number, Set<(id: number, account: number, ...args: any[]) => void>> = new Map();

  // Telegram Event IDs
  public static readonly dialogsNeedReload: number = 1;
  public static readonly updateInterfaces: number = 2;
  public static readonly messagesDidLoad: number = 3;
  public static readonly didLoadedReplyMessages: number = 4;
  public static readonly userFullInfoDidLoad: number = 5;
  public static readonly chatInfoDidLoad: number = 6;
  public static readonly encryptedChatCreated: number = 7;
  public static readonly dialogPhotosLoaded: number = 8;
  public static readonly notificationsSettingsUpdated: number = 9;
  public static readonly didReceivedNewMessages: number = 10;
  public static readonly messagesDeleted: number = 11;
  public static readonly messageReceivedByAck: number = 12;
  public static readonly messageSendError: number = 13;
  public static readonly chatDidCreated: number = 14;
  public static readonly dialogDeleted: number = 15;
  public static readonly totalUnreadPushedOnce: number = 16;
  public static readonly mainUserInfoChanged: number = 17;
  public static readonly privacyRulesUpdated: number = 18;
  public static readonly proxySettingsChanged: number = 19;
  public static readonly appDidLogout: number = 20;

  private constructor() {}

  public static getInstance(): NotificationCenter {
    if (!NotificationCenter.instance) {
      NotificationCenter.instance = new NotificationCenter();
    }
    return NotificationCenter.instance;
  }

  public static getGlobalInstance(): NotificationCenter {
    return NotificationCenter.getInstance();
  }

  public addObserver(observer: (id: number, account: number, ...args: any[]) => void, id: number): void {
    let set = this.observers.get(id);
    if (!set) {
      set = new Set();
      this.observers.set(id, set);
    }
    set.add(observer);
  }

  public removeObserver(observer: (id: number, account: number, ...args: any[]) => void, id: number): void {
    const set = this.observers.get(id);
    if (set) {
      set.delete(observer);
      if (set.size === 0) {
        this.observers.delete(id);
      }
    }
  }

  public postNotificationName(id: number, ...args: any[]): void {
    const set = this.observers.get(id);
    if (set && set.size > 0) {
      const copy = Array.from(set);
      copy.forEach((observer) => {
        try {
          observer(id, 0, ...args);
        } catch (e) {
          console.error(`[NotificationCenter] Error in observer for event ${id}:`, e);
        }
      });
    }
  }

  public hasObservers(id: number): boolean {
    const set = this.observers.get(id);
    return Boolean(set && set.size > 0);
  }
}

File: TMessagesProject/src/main/java/org/telegram/messenger/NotificationCenter.java
package org.telegram.messenger;

import android.os.Handler;
import android.os.Looper;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * NotificationCenter - Replicates DrKLO/Telegram NotificationCenter.java
 * Handles observer patterns for UI synchronization and multi-account state updates.
 */
public class NotificationCenter {
    public static final int didReceivedNewMessages = 1;
    public static final int updateInterfaces = 2;
    public static final int dialogsNeedReload = 3;
    public static final int messagesDeleted = 4;
    public static final int mainUserInfoChanged = 5;
    public static final int appDidLogout = 6;
    public static final int didUpdateConnectionState = 7;
    public static final int accountSwitched = 8;
    public static final int updatesDidReceived = 9;

    public interface NotificationCenterDelegate {
        void didReceivedNotification(int id, int account, Object... args);
    }

    private static final NotificationCenter[] Instance = new NotificationCenter[UserConfig.MAX_ACCOUNT_COUNT];
    private static NotificationCenter globalInstance;

    private final int currentAccount;
    private final HashMap<Integer, ArrayList<NotificationCenterDelegate>> observers = new HashMap<>();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    public static NotificationCenter getInstance(int num) {
        if (num < 0 || num >= UserConfig.MAX_ACCOUNT_COUNT) {
            num = 0;
        }
        NotificationCenter local = Instance[num];
        if (local == null) {
            synchronized (NotificationCenter.class) {
                local = Instance[num];
                if (local == null) {
                    Instance[num] = local = new NotificationCenter(num);
                }
            }
        }
        return local;
    }

    public static NotificationCenter getGlobalInstance() {
        if (globalInstance == null) {
            synchronized (NotificationCenter.class) {
                if (globalInstance == null) {
                    globalInstance = new NotificationCenter(-1);
                }
            }
        }
        return globalInstance;
    }

    private NotificationCenter(int account) {
        currentAccount = account;
    }

    public void addObserver(NotificationCenterDelegate observer, int id) {
        ArrayList<NotificationCenterDelegate> list = observers.computeIfAbsent(id, k -> new ArrayList<>());
        if (!list.contains(observer)) {
            list.add(observer);
        }
    }

    public void removeObserver(NotificationCenterDelegate observer, int id) {
        ArrayList<NotificationCenterDelegate> list = observers.get(id);
        if (list != null) {
            list.remove(observer);
        }
    }

    public void postNotificationName(int id, Object... args) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            postNotificationNameInternal(id, args);
        } else {
            mainHandler.post(() -> postNotificationNameInternal(id, args));
        }
    }

    private void postNotificationNameInternal(int id, Object... args) {
        ArrayList<NotificationCenterDelegate> list = observers.get(id);
        if (list != null && !list.isEmpty()) {
            ArrayList<NotificationCenterDelegate> copy = new ArrayList<>(list);
            for (NotificationCenterDelegate delegate : copy) {
                delegate.didReceivedNotification(id, currentAccount, args);
            }
        }
    }
}
